### PR TITLE
Fix: check validity before setting the default value

### LINKED ISSUE #6086: [BUG] `checkValue` does not work in `RapidsConf`
**Describe the bug**
In `RapidsConf.scala`, there are some `ConfigBuiler` object. For example,
```scala
  val MULTITHREAD_READ_NUM_THREADS = conf("spark.rapids.sql.multiThreadedRead.numThreads")
      .doc("The maximum number of threads on each executor to use for reading small " +
        "files in parallel. This can not be changed at runtime after the executor has " +
        ...
        ".")
      .integerConf
      .checkValue(v => v > 0, "The thread count must be greater than zero.")
      .createWithDefault(0)
```
Here `checkValue` does not work. `createWithDefault(-1)` or `createWithDefault(0)` will not throw any exception.

Even if `"spark.rapids.sql.multiThreadedRead.numThreads" = -1` is set explicitly by user, there is no exception, either.

---

**Steps/Code to reproduce bug**
Run spark-rapids locally via:
```shell
spark-shell --jars ${RAPIDS_JAR} \
  --conf spark.sql.adaptive.enabled=false \
  --conf spark.plugins=com.nvidia.spark.SQLPlugin \
  --conf spark.rapids.sql.multiThreadedRead.numThreads=-1
 
scala> val key = "spark.rapids.sql.multiThreadedRead.numThreads"                                                                    
key: String = spark.rapids.sql.multiThreadedRead.numThreads                                                                         
                                                                                                                                    
scala> spark.conf.get(key)                                                                                                          
res0: String = -1                                                                                                                   
                                                                                                                                    
scala> :quit
```


**Expected behavior**
When spark.rapids.sql.multiThreadedRead.numThreads <= 0, we should throw an `IllegalArgument` exception.

In original spark, if we set an invalid parameter, there is an `IllegalArgument`.

```shell
$ spark-shell --conf spark.sql.shuffle.partitions=-1                                                                                                                              

scala> spark.conf.get("spark.sql.shuffle.partitions")
java.lang.IllegalArgumentException: Error while instantiating 'org.apache.spark.sql.hive.HiveSessionStateBuilder':
  at org.apache.spark.sql.SparkSession$.org$apache$spark$sql$SparkSession$$instantiateSessionState(SparkSession.scala:1178)
  at org.apache.spark.sql.SparkSession.$anonfun$sessionState$2(SparkSession.scala:162)
  at scala.Option.getOrElse(Option.scala:189)
  at org.apache.spark.sql.SparkSession.sessionState$lzycompute(SparkSession.scala:160)
  at org.apache.spark.sql.SparkSession.sessionState(SparkSession.scala:157)
  at org.apache.spark.sql.SparkSession.conf$lzycompute(SparkSession.scala:185)
  at org.apache.spark.sql.SparkSession.conf(SparkSession.scala:185)
  ... 47 elided
Caused by: java.lang.IllegalArgumentException: '-1' in spark.sql.shuffle.partitions is invalid. The value of spark.sql.shuffle.partitions must be positive
...

```

### PR DESCRIPTION
**Pull request by sinkinben**
Close #6086.

For each item in RapidsConf, we should execute `checkValue` before assigning it a default value.

The `converter` is wrapped with `checkValue`. It is similar to spark.

```scala
  // See ConfigBuilder.scala
  /** Creates a [[ConfigEntry]] that has a default value. */
  def createWithDefault(default: T): ConfigEntry[T] = {
    // Treat "String" as a special case, so that both createWithDefault and createWithDefaultString
    // behave the same w.r.t. variable expansion of default values.
    default match {
      case str: String => createWithDefaultString(str)
      case _ =>
        val transformedDefault = converter(stringConverter(default))
        val entry = new ConfigEntryWithDefault[T](parent.key, parent._prependedKey,
          parent._prependSeparator, parent._alternatives, transformedDefault, converter,
          stringConverter, parent._doc, parent._public, parent._version)
        parent._onCreate.foreach(_ (entry))
        entry
    }
  }
```

### CODE REVIEW

**Comment by sinkinben** [comment]
But there is still one issue.

For example, if we set `--conf spark.rapids.sql.multiThreadedRead.numThreads=-1` in the command line, `checkValue` still is not executed.

This problem confuses me, I have no idea how to fix it in an elegant way.

In spark, when calling `spark.conf.get()`, it will check its validity.

```shell
$ spark-shell --conf spark.sql.shuffle.partitions=-1                                                                                                                              

scala> spark.conf.get("spark.sql.shuffle.partitions")
java.lang.IllegalArgumentException: Error while instantiating 'org.apache.spark.sql.hive.HiveSessionStateBuilder':
  at org.apache.spark.sql.SparkSession$.org$apache$spark$sql$SparkSession$$instantiateSessionState(SparkSession.scala:1178)
  at org.apache.spark.sql.SparkSession.$anonfun$sessionState$2(SparkSession.scala:162)
  at scala.Option.getOrElse(Option.scala:189)
  at org.apache.spark.sql.SparkSession.sessionState$lzycompute(SparkSession.scala:160)
  at org.apache.spark.sql.SparkSession.sessionState(SparkSession.scala:157)
  at org.apache.spark.sql.SparkSession.conf$lzycompute(SparkSession.scala:185)
  at org.apache.spark.sql.SparkSession.conf(SparkSession.scala:185)
  ... 47 elided
Caused by: java.lang.IllegalArgumentException: '-1' in spark.sql.shuffle.partitions is invalid. The value of spark.sql.shuffle.partitions must be positive
...
```

**Comment by sinkinben** [comment]
Check the validity of config value set by users.

```scala
  // In class RapidsConf(conf: Map[String, String])...
  def get[T](entry: ConfEntry[T]): T = {
    // 'converter' is wrapped with 'checkValue' (if the config item has), so we can check the
    //  validity of the value set by user.
    val ret = entry.get(conf)
    if (ret != None) entry.converter(ret.toString)
    ret
  }
```

Since the config items in class `RapidsConf` are lazy values, **only when they are used for 1st time**, the plugin will call the method above to get and check the value.

---
For example, if we assign 2 to `spark.rapids.memory.gpu.maxAllocFraction` (which should be in `[0, 1]`), then we throw exception.
```shell
~$ spark-shell --jars ${RAPIDS_JAR}   \
--conf spark.sql.adaptive.enabled=false   \
--conf spark.plugins=com.nvidia.spark.SQLPlugin   \
--conf spark.rapids.memory.gpu.maxAllocFraction=2        
                                                                                                                                     
...
               
22/07/28 11:41:03 ERROR RapidsExecutorPlugin: Exception in the executor plugin, shutting down!                                                                                                        
java.lang.IllegalArgumentException: The fraction value must be in [0, 1].                                                                                                                             
        at com.nvidia.spark.rapids.TypedConfBuilder.$anonfun$checkValue$1(RapidsConf.scala:210)                                                                                                       
        at com.nvidia.spark.rapids.TypedConfBuilder.$anonfun$transform$1(RapidsConf.scala:203)                                                                                                        
        at scala.Option.map(Option.scala:230)                                                               
```

---

However, there is still one issue when the config item is never used.

```shell
spark-shell --jars ${RAPIDS_JAR}   \
--conf spark.sql.adaptive.enabled=false   \
--conf spark.plugins=com.nvidia.spark.SQLPlugin   \
--conf spark.rapids.sql.multiThreadedRead.numThreads=-1

scala> val key = "spark.rapids.sql.multiThreadedRead.numThreads"
key: String = spark.rapids.sql.multiThreadedRead.numThreads

scala> spark.conf.get(key)
res0: String = -1
```

Since there is no code to touch `RapidsConf.multiThreadReadNumThreads`, it will not execute the `get` method above.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 1]
```diff
@@ -1620,7 +1623,11 @@ class RapidsConf(conf: Map[String, String]) extends Logging {
   }
 
   def get[T](entry: ConfEntry[T]): T = {
-    entry.get(conf)
+    // 'converter' is wrapped with 'checkValue' (if the config item has), so we can check the
```
Comment (firestarman) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:1626]: Do not do this here, let the `entry` take care of the converting things in its `get`.

Comment (sinkinben) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala]: Fixed.

**Comment by sinkinben** [comment]
build

**Review by firestarman — APPROVED** [review]

**Comment by firestarman** [comment]
build

**Pull request merged**